### DISCUSSION TITLE
How to correctly return errors and partial data in list query

### DISCUSSION THREAD

**Discussion by tsteenkamp** [MichalLytek/type-graphql#1161] [Q&A]
Hi,

so I have a root query that receives a list of IDs of entities and should return details about those or an error if it cannot find the corresponding entity. For example:

```
@ArgsType()
export class ProductsArgs {
  @Field(() => [Number])
  ids: number[]
}

@ObjectType()
export class Product {
  @Field(() => Number)
  id: number
}

const products = [1, 3].map(id => {
  const product = new Product()
  product.id = id
  return product
})

@Resolver()
export class ProductsResolver {
  @Query(() => [Product])
  async products(@Args() { ids }: ProductsArgs): Promise<Array<Product | GraphQLError>> {
    const list = ids.map(id => products.find(f => f.id === id) || new GraphQLError(`Product ${id} not found`))
    console.log(`list`, list)
    return list
  }
}
```

If I query now something like:
```
query {
  products(ids: [1, 2, 3]) {
    id
  }
}
```

I can see in stdout the expected list of two Products and one GraphQLError, but the graphql response object only contains the errors map, not the valid partial data. 

Is there a way to return partial valid data in this case? I'm pretty sure I have seen other / more deeply nested requests, where type-graphql handled partial data and errors as I expected it to do. I was surprised that this was handled as if the error was thrown, altough it is "only" mapped into the result set.

**Comment by MichalLytek** [comment]
You need to throw the error, not return the error.

```ts
const list = ids.map(id => products.find(product => product.id === id));
if (list.some(item => item === undefined)) {
  throw new Error("...");
}
```

**Reply by tsteenkamp** [reply]
Throwing the error instead of returning (and thus mapping it to the result set) is even worse, because it will exit the current code block and all valid results are lost - or am I missing something?

The response I'm aiming for is this:

```
{
  "errors": [
    {
      "message": "Product 2 not found",
      "locations": [],
      "path": [
        "products",
        1,
      ],
      "extensions": {
        "code": "NOT_FOUND"
      }
    }
  ],
  "data": {
    "products": [
      {
        "id": 1
      },
      {
        "id": 3
      },
    ]
  }
}
```

The reason I'm asking is that I want to implement a query required for schema stitching. So the downstream service (written in type-graphql) has to implement a query that is used by the gateway to query additional information for types where different sub-services contribute to.

Specifically what I'm trying to do is explained here: 
https://github.com/gmac/schema-stitching-handbook/tree/master/type-merging-arrays#error-handling
and here:
https://github.com/gmac/schema-stitching-handbook/blob/master/type-merging-arrays/services/manufacturers/schema.js

**Reply by MichalLytek** [reply]
Not possible in GraphQL that single query both resolves data and throw error. 

Maybe try to use custom response object type with `data` and `errors` properties?

**Reply by tsteenkamp** [reply]
It is definitely possible. See a minimal example here: https://stackblitz.com/edit/graphql-lxenm4

**Reply by MichalLytek** [reply]
```
{
  "errors": [
    {
      "message": "Manufacturer '3' not found.",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "manufacturers",
        1
      ]
    }
  ],
  "data": {
    "manufacturers": [
      {
        "id": "1",
        "name": "Apple"
      },
      null,
      {
        "id": "2",
        "name": "Macmillan"
      }
    ]
  }
}
```

TypeGraphQL does not add any logic for resolving values, so maybe the problem here is with exposing schema by apollo or something?

**Reply by MichalLytek** [reply]
```
 @Query(() => [Product])
 ```
 ```
  manufacturers(ids: [ID!]!): [Manufacturer]!
  ```
  
  Your TypeGraphQL type is not nullable

**Reply by tsteenkamp** [reply]
Wow, thanks a lot! Stupid mistake.
